File: frontend/src/api/deals.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import { apiClient } from './client'

// ── Types ─────────────────────────────────────────────────────────────────────

export interface PipelineStageOut {
    id: string
    name: string
    order: number
    default_probability: number
    is_closed: boolean
    is_won: boolean
    color: string | null
}

export interface DealOut {
    id: string
    title: string
    amount: string
    currency: string
    probability: number
    stage: string
    expected_close: string | null
    origin: string | null
    notes: string | null
    tags: string[]
    is_locked: boolean
    weighted_amount: string
    company_id: string | null
    contact_id: string | null
    company_name: string | null
    contact_name: string | null
    has_project: boolean
    created_at: string
    updated_at: string
}

export interface DealCreate {
    title: string
    amount?: string | number
    currency?: string
    probability?: number
    stage?: string
    expected_close?: string | null
    origin?: string | null
    notes?: string | null
    tags?: string[]
    company_id?: string | null
    contact_id?: string | null
}

export interface DealPatch {
    title?: string
    amount?: string | number
    currency?: string
    probability?: number
    stage?: string
    expected_close?: string | null
    origin?: string | null
    notes?: string | null
    tags?: string[]
    company_id?: string | null
    contact_id?: string | null
}

export interface DealMove {
    stage: string
}

export interface DealList {
    items: DealOut[]
    total: number
    page: number
    page_size: number
}

export interface DealListParams {
    stage?: string
    close_before?: string
    company_id?: string
    page?: number
    page_size?: number
}

// ── Query keys ────────────────────────────────────────────────────────────────

export const dealKeys = {
    all: ['deals'] as const,
    list: (params?: DealListParams) => [...dealKeys.all, 'list', params] as const,
    detail: (id: string) => [...dealKeys.all, 'detail', id] as const,
    stages: () => ['pipeline', 'stages'] as const,
}

// ── API calls ─────────────────────────────────────────────────────────────────

export const dealsApi = {
    list: (params?: DealListParams) =>
        apiClient.get<DealList>('/deals', { params }).then((r) => r.data),
    get: (id: string) => apiClient.get<DealOut>(`/deals/${id}`).then((r) => r.data),
    create: (data: DealCreate) => apiClient.post<DealOut>('/deals', data).then((r) => r.data),
    patch: (id: string, data: DealPatch) =>
        apiClient.patch<DealOut>(`/deals/${id}`, data).then((r) => r.data),
    move: (id: string, data: DealMove) =>
        apiClient.post<DealOut>(`/deals/${id}/move`, data).then((r) => r.data),
    delete: (id: string) => apiClient.delete(`/deals/${id}`).then((r) => r.data),
    getStages: () =>
        apiClient.get<PipelineStageOut[]>('/pipeline/stages').then((r) => r.data),
    updateStages: (stages: Omit<PipelineStageOut, 'id' | 'created_at' | 'updated_at'>[]) =>
        apiClient.put<PipelineStageOut[]>('/pipeline/stages', { stages }).then((r) => r.data),
}

// ── Hooks ─────────────────────────────────────────────────────────────────────

export function useDeals(params?: DealListParams) {
    return useQuery({
        queryKey: dealKeys.list(params),
        queryFn: () => dealsApi.list(params),
    })
}

export function useDeal(id: string) {
    return useQuery({
        queryKey: dealKeys.detail(id),
        queryFn: () => dealsApi.get(id),
        enabled: Boolean(id),
    })
}

export function usePipelineStages() {
    return useQuery({
        queryKey: dealKeys.stages(),
        queryFn: dealsApi.getStages,
        staleTime: 5 * 60 * 1000, // 5 min — les stages changent rarement
    })
}

export function useCreateDeal() {
    const qc = useQueryClient()
    return useMutation({
        mutationFn: (data: DealCreate) => dealsApi.create(data),
        onSuccess: () => qc.invalidateQueries({ queryKey: dealKeys.all }),
    })
}

export function usePatchDeal() {
    const qc = useQueryClient()
    return useMutation({
        mutationFn: ({ id, data }: { id: string; data: DealPatch }) => dealsApi.patch(id, data),
        onSuccess: (_, { id }) => {
            qc.invalidateQueries({ queryKey: dealKeys.detail(id) })
            qc.invalidateQueries({ queryKey: dealKeys.all })
        },
    })
}

export function useMoveDeal() {
    const qc = useQueryClient()
    return useMutation({
        mutationFn: ({ id, stage }: { id: string; stage: string }) =>
            dealsApi.move(id, { stage }),
        onSuccess: () => qc.invalidateQueries({ queryKey: dealKeys.all }),
    })
}

export function useDeleteDeal() {
    const qc = useQueryClient()
    return useMutation({
        mutationFn: (id: string) => dealsApi.delete(id),
        onSuccess: () => qc.invalidateQueries({ queryKey: dealKeys.all }),
    })
}

File: backend/app/schemas/deals.py
"""Schémas Pydantic — Deals (Opportunités)."""
from __future__ import annotations

import uuid
from datetime import date, datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, Field

DealStage = Literal[
    "Découverte", "Qualification", "Proposition", "Négociation", "Gagné", "Perdu"
]


class DealCreate(BaseModel):
    title: str = Field(min_length=1, max_length=300)
    amount: Decimal = Field(default=Decimal("0"), ge=0)
    currency: str = Field(default="EUR", max_length=3)
    probability: int = Field(default=0, ge=0, le=100)
    stage: str = Field(default="Découverte", max_length=50)
    expected_close: date | None = None
    origin: str | None = Field(default=None, max_length=100)
    notes: str | None = Field(default=None, max_length=10000)
    tags: list[str] = Field(default_factory=list)
    company_id: uuid.UUID | None = None
    contact_id: uuid.UUID | None = None


class DealPatch(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=300)
    amount: Decimal | None = Field(default=None, ge=0)
    currency: str | None = Field(default=None, max_length=3)
    probability: int | None = Field(default=None, ge=0, le=100)
    stage: str | None = Field(default=None, max_length=50)
    expected_close: date | None = None
    origin: str | None = Field(default=None, max_length=100)
    notes: str | None = Field(default=None, max_length=10000)
    tags: list[str] | None = None
    company_id: uuid.UUID | None = None
    contact_id: uuid.UUID | None = None


class DealMove(BaseModel):
    """Déplacer un deal vers un nouveau stage (kanban)."""
    stage: str = Field(min_length=1, max_length=50)


class DealOut(BaseModel):
    id: uuid.UUID
    title: str
    amount: Decimal
    currency: str
    probability: int
    stage: str
    expected_close: date | None
    origin: str | None
    notes: str | None
    tags: list[str]
    is_locked: bool
    weighted_amount: Decimal
    company_id: uuid.UUID | None
    contact_id: uuid.UUID | None
    created_at: datetime
    updated_at: datetime
    # Champs enrichis (join)
    company_name: str | None = None
    contact_name: str | None = None
    has_project: bool = False

    model_config = {"from_attributes": True}


class DealList(BaseModel):
    items: list[DealOut]
    total: int
    page: int
    page_size: int
